File: firefly/src/main/java/com/firefly/server/http2/router/impl/AbstractRegexMatcher.java
package com.firefly.server.http2.router.impl;

import com.firefly.server.http2.router.Matcher;
import com.firefly.server.http2.router.Router;

import java.util.*;
import java.util.regex.Pattern;

/**
 * @author Pengtao Qiu
 */
abstract public class AbstractRegexMatcher implements Matcher {

    protected Map<RegexRule, Set<Router>> regexMap;

    protected static class RegexRule {
        final String rule;
        final Pattern pattern;

        protected RegexRule(String rule) {
            this.rule = rule;
            pattern = Pattern.compile(rule);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            RegexRule regexRule = (RegexRule) o;
            return Objects.equals(rule, regexRule.rule);
        }

        @Override
        public int hashCode() {
            return Objects.hash(rule);
        }
    }

    protected Map<RegexRule, Set<Router>> regexMap() {
        if (regexMap == null) {
            regexMap = new HashMap<>();
        }
        return regexMap;
    }

    @Override
    public void add(String rule, Router router) {
        regexMap().computeIfAbsent(new RegexRule(rule), k -> new HashSet<>()).add(router);
    }

    @Override
    public MatchResult match(String value) {
        if (regexMap == null) {
            return null;
        }

        Set<Router> routers = new HashSet<>();
        Map<Router, Map<String, String>> parameters = new HashMap<>();

        regexMap.entrySet()
                .forEach(e -> {
                    java.util.regex.Matcher m = e.getKey().pattern.matcher(value);
                    if (m.matches()) {
                        routers.addAll(e.getValue());
                        m = e.getKey().pattern.matcher(value);

                        Map<String, String> param = new HashMap<>();
                        while (m.find()) {
                            for (int i = 1; i <= m.groupCount(); i++) {
                                param.put("group" + i, m.group(i));
                            }
                        }
                        if (!param.isEmpty()) {
                            e.getValue().forEach(router -> parameters.put(router, param));
                        }
                    }
                });
        if (routers.isEmpty()) {
            return null;
        } else {
            return new MatchResult(routers, parameters, getMatchType());
        }
    }
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/AbstractRegexMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Router
import java.util.*
import java.util.regex.Pattern


abstract class AbstractRegexMatcher :AbstractMatcher<AbstractRegexMatcher.RegexRule>(), Matcher {

    companion object {
        const val paramName = "group"
    }

    class RegexRule(val rule: String) {
        val pattern: Pattern = Pattern.compile(rule)

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as RegexRule
            return rule == other.rule
        }

        override fun hashCode(): Int {
            return rule.hashCode()
        }
    }

    override fun add(rule: String, router: Router) {
        routersMap.computeIfAbsent(RegexRule(rule)) { TreeSet() }.add(router)
    }

    override fun match(value: String): Matcher.MatchResult? {
        if (routersMap.isEmpty()) return null

        val routers = TreeSet<Router>()
        val parameters = HashMap<Router, Map<String, String>>()

        routersMap.forEach { (rule, routerSet) ->
            var matcher = rule.pattern.matcher(value)
            if (matcher.matches()) {
                routers.addAll(routerSet)
                matcher = rule.pattern.matcher(value)
                val param: MutableMap<String, String> = HashMap()
                while (matcher.find()) {
                    for (i in 1..matcher.groupCount()) {
                        param["$paramName$i"] = matcher.group(i)
                    }
                }
                if (param.isNotEmpty()) {
                    routerSet.forEach { router -> parameters[router] = param }
                }
            }
        }

        return if (routers.isEmpty()) null else Matcher.MatchResult(routers, parameters, matchType)
    }
}